test/Quake/cast_fold.qke

```diff
@@ -0,0 +1,122 @@
+
+func.func @b1() -> f32 {
+  %0 = arith.constant 22 : i64
+  %1 = cc.cast signed %0 : (i64) -> f32
+  return %1 : f32
+}
+
+func.func @b2() -> f32 {
+  %0 = arith.constant -21 : i64
+  %1 = cc.cast unsigned %0 : (i64) -> f32
+  return %1 : f32
+}
+
+func.func @b3() -> f32 {
+  %0 = arith.constant 2.34 : f64
+  %1 = cc.cast %0 : (f64) -> f32
+  return %1 : f32
+}
+
+func.func @b4() -> f64 {
+  %0 = arith.constant 6.34 : f32
+  %1 = cc.cast %0 : (f32) -> f64
+  return %1 : f64
+}
+
+func.func @b5() -> i32 {
+  %0 = arith.constant 94 : i64
+  %1 = cc.cast signed %0 : (i64) -> i32
+  return %1 : i32
+}
+
+func.func @b6() -> i64 {
+  %0 = arith.constant 73 : i16
+  %1 = cc.cast signed %0 : (i16) -> i64
+  return %1 : i64
+}
+
+func.func @b7() -> i64 {
+  %0 = arith.constant -82 : i16
+  %1 = cc.cast unsigned %0 : (i16) -> i64
+  return %1 : i64
+}
+
+func.func @b8() -> i64 {
+  %0 = arith.constant -6.29 : f32
+  %1 = cc.cast signed %0 : (f32) -> i64
+  return %1 : i64
+}
+
+func.func @b9() -> i32 {
+  %0 = arith.constant 7.11 : f64
+  %1 = cc.cast unsigned %0 : (f64) -> i32
+  return %1 : i32
+}
+
+func.func @c0() -> i32 {
+  %0 = arith.constant -7.11 : f64
+  %1 = cc.cast unsigned %0 : (f64) -> i32
+  return %1 : i32
+}
+
+// CHECK-LABEL:   func.func @b1() -> f32 {
+// CHECK:           %[[VAL_0:.*]] = arith.constant 2.200000e+01 : f32
+// CHECK:           return %[[VAL_0]] : f32
+// CHECK:         }
+
+// CHECK-LABEL:   func.func @b2() -> f32 {
+// CHECK:           %[[VAL_0:.*]] = arith.constant 1.84467441E+19 : f32
+// CHECK:           return %[[VAL_0]] : f32
+// CHECK:         }
+
+// CHECK-LABEL:   func.func @b3() -> f32 {
+// CHECK:           %[[VAL_0:.*]] = arith.constant 2.340000e+00 : f32
+// CHECK:           return %[[VAL_0]] : f32
+// CHECK:         }
+
+// CHECK-LABEL:   func.func @b4() -> f64 {
+// CHECK:           %[[VAL_0:.*]] = arith.constant 6.3400001525878906 : f64
+// CHECK:           return %[[VAL_0]] : f64
+// CHECK:         }
+
+// CHECK-LABEL:   func.func @b5() -> i32 {
+// CHECK:           %[[VAL_0:.*]] = arith.constant 94 : i32
+// CHECK:           return %[[VAL_0]] : i32
+// CHECK:         }
+
+// CHECK-LABEL:   func.func @b6() -> i64 {
+// CHECK:           %[[VAL_0:.*]] = arith.constant 73 : i64
+// CHECK:           return %[[VAL_0]] : i64
+// CHECK:         }
+
+// CHECK-LABEL:   func.func @b7() -> i64 {
+// CHECK:           %[[VAL_0:.*]] = arith.constant -82 : i64
+// CHECK:           return %[[VAL_0]] : i64
+// CHECK:         }
+
+// CHECK-LABEL:   func.func @b8() -> i64 {
+// CHECK:           %[[VAL_0:.*]] = arith.constant -6 : i64
+// CHECK:           return %[[VAL_0]] : i64
+// CHECK:         }
+
+// CHECK-LABEL:   func.func @b9() -> i32 {
+// CHECK:           %[[VAL_0:.*]] = arith.constant 7 : i32
+// CHECK:           return %[[VAL_0]] : i32
+// CHECK:         }
+
+// On amd64, the result is 4294967289 (-7) but on arm64 it folds to 0.
+// I've verified this behavior on an arm64 machine with g++ and clang++.
```

Comment:
Ouch. I don't know what the precedence is here, but as a user, I'd rather it throw an error than produce platform-specific behavior. If this is really used for constant folding when all the values are concrete values, would it be possible to throw an error or simply not fold it instead?

Reply:
We're folding this exactly like the C++ compilers do on those systems. That's not a bug. It's a bit strange that it's inconsistent, I agree, but this is the correct thing to do and has been verified.

Reply:
`clang++` also supports `-fsanitize=undefined` to notify the user if they encounter this behavior. Would that flag be honored if/when running it in Quake, too?

E.g.
```cpp
int main(int argc, char *argv[]) {
  double x = -7.11;
  unsigned y = static_cast<unsigned>(x);
  printf("x = %f   y = %u\n", x, y);
  return 0;
}
```
```bash
$ clang++ -fsanitize=undefined test.cpp
$ ./a.out
test.cpp:14:38: runtime error: -7.11 is outside the range of representable values of type 'unsigned int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior test.cpp:14:38 in 
x = -7.110000   y = 4294967289
```

Reply:
This isn't `constexpr`.

What you're suggesting here is that library mode, AOT compilation, and JIT compilation all work differently. That is, identical code that runs in library mode, compiles with the AOT compiler without C++ warnings, will crash/raise an error in the JIT compiler. Worse, the only reason the JIT compile would fail is because I added constant folding earlier in the pipeline.

While I appreciate that undefined behavior is something the user may not expect, it may also be something they fully expect because they've read the standard and know their platform.

Reply:
> What you're suggesting here is that library mode, AOT compilation, and JIT compilation all work differently.

Yes, that's a good point, but I think it's an acceptable deviation in this case because we are informing the user about their use of unsafe code. Users should not be writing code that makes ~~architecture-dependent~~ erroneous assumptions [*], and if we have the opportunity to warn them about it, we should. If we had the ability to do this in library mode, I would want us to do that, too, but I don't think the fact that we can't do it in library mode should prevent us from helping the user out in JIT compilation.

[*] I just discovered that even without `constexpr`, the results of the above test program change based on optimization level (on amd64).

| Compiler | -O0 | -O1 | -O2 | -O3 |
| --- | --- | --- | --- | --- |
| g++ | 4,294,967,289 | 0 | 0 | 0 |
| clang++ | 4,294,967,289 | 3,583,640,872 | 3,841,045,096 | 2,228,773,352 |

I think that nearly every user would prefer to be notified about that bug in their code if the compiler is able to conclusively determine that they are doing things incorrectly, and in this case, I would argue that we can conclusively determine if they are doing something wrong for the price of a single if check.

That's just my two cents.

Reply:
If the programmer writes this code
```c++
   double d = ...;
   unsigned u = (unsigned)d;
```
which is standard conforming code, then we will produce correct results where this is well-defined and results 100% consistent with the C++ compiler and library mode otherwise. I disagree that heroic efforts to sanitize the floating-point values is necessary.

Reply:
Some other cast situations that may do something unexpected include
```c++
int i = ...;
unsigned u = i;

char ch = ...;
short sh = ch;

double d = ....;
float f = d;

long ll = ...;
int i = ll;

char *cp = ...;
*(int*)cp = 42;
```
I don't think nvq++ should be trying to warn about the conditions and ways these might go wrong.

Reply:
Unless I'm mistaken, I believe all of those examples follow well-defined behaviors with the possible exception of the pointer handling. The pointer example would not be handled by this pass because the pointers are not constant values anyway. The `char` example highlights that some compilers choose signed char while others choose unsigned char by default, but for any given compiler, there *is* a default sign value, and the spec says what should happen depending on what the sign of the char is, so that results in well-defined behavior (unlike the example that started this thread).

I disagree that if checks against constant values are heroic efforts. If they were, I suspect both clang++ and g++ wouldn't have bothered with checking for this exact error condition when handling `constexpr` items, but they clearly do; it's trivially easy to do when you're working with constants.

At a minimum, it appears the comment about "On amd64, the result is 4294967289 (-7)" isn't strictly correct because the actual values depend on host platform, host compiler, and compilation optimization settings, and perhaps what the phase of the moon is 😄.

If it's more convenient, I can write an issue and circle back to this whole thing later.

Reply:
#1959 written as an enhancement issue.